### PR TITLE
Make autobump pr creation safer

### PR DESCRIPTION
**Pull request by dcsobral**
The PR creation task is now responsible for handling existing PRs check
and PR updates. Pull request and commit messages were improved, as well
as error messages.

These changes required extending github4s to add draft PR support, and
implement a number of methods that were not available. It also added
support to github's v4 API, which is graphql-based.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -6,12 +6,14 @@ ThisBuild / scalaVersion := "2.12.10"
...
   .aggregate(core, artifact, plugin)
+  .settings(name := "sbt-precog-root")
   .settings(noPublishSettings)
 
 lazy val core = project.in(file("core"))
   .settings(
     name := "sbt-precog-core",
-    scalacStrictMode := false)
+    scalacStrictMode := false,
+    /*Compile/sourceGenerators += graphqlCodegen.value*/)
```
Comment (dcsobral) [inline build.sbt]: Remove this.

File: .gitignore [hunk 2]
```diff
@@ -6,6 +6,8 @@ tags
 
 /core/src/main/resources/core/common-secrets.yml
 
+.graphqlconfig
```
Comment (dcsobral) [inline .gitignore:9]: This is a configuration file for an IntelliJ plugin/node app. It contains the authorization token, which is why I'm forcibly ignoring it.

File: core/build.sbt [hunk 3]
```diff
@@ -3,8 +3,17 @@ libraryDependencies += "org.yaml" % "snakeyaml" % "1.26"
 addSbtPlugin("io.crashbox"       % "sbt-gpg"            % "0.2.1")
 addSbtPlugin("com.codecommit"    % "sbt-github-actions" % "0.1-58d9629")
 addSbtPlugin("de.heikoseeberger" % "sbt-header"         % "5.4.0")
-addSbtPlugin("com.dcsobral"      % "sbt-trickle"        % "0.2.4")
+addSbtPlugin("com.dcsobral"      % "sbt-trickle"        % "0.3-37c1a2c")
```
Comment (dcsobral) [inline core/build.sbt:6]: I'm waiting to be certain of what the API needs to look like before I make a release.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+    }
+
+    val app = for {
+      (dir, branchName, maybePullRequest, changes, label) <- EitherT(tryUpdateDependencies)
+      _ <- EitherT(tryCommit(dir))
+      _ <- EitherT(tryPush(dir, branchName))
+      pullRequest <- EitherT(createOrUpdatePullRequest(branchName, changes, label, maybePullRequest))
+    } yield pullRequest
+
+    app.leftSemiflatMap(_.warn(log)).value.unsafeRunSync().isRight
```
Comment (dcsobral) [inline core/src/main/scala/slamdata/AutoBump.scala]: I'm considering returning `IO` here, to make it testable. Thoughts?

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+        _ <- run(s"git checkout $flag $branchName", log, workingDir = Some(dirFile))
+        lines <- run(s"$sbt trickleUpdateDependencies", log, workingDir = Some(dirFile))
+        changes = extractChanges(lines)
+        label = extractLabel(lines)
+        _ <- run(s"$sbt trickleIsUpToDate", log, workingDir = Some(dirFile))
+        updateResult <- run(s"$sbt update", log, merge = true, Some(dirFile)).attempt
+      } yield updateResult.bimap(_ => UpdateError, _ => (dirFile, branchName, oldestPullRequest, changes, label))
+    }
+
+    // TODO: add changes to commit message?
```
Comment (dcsobral) [inline core/src/main/scala/slamdata/AutoBump.scala]: @djspiewak Feedback on this, please.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    } yield pullRequest.asRight[Warnings]
+
+    def ifNotOldest(oldest: PullRequestDraft, pullRequest: PullRequestDraft): IO[Either[Warnings, PullRequestDraft]] = for {
+      _ <- close(pullRequest, s"${pullRequest.title} (preceded by #${oldest.number})")
+      _ <- IO(log.info(s"Closed $owner/$repoSlug#${pullRequest.number}, preceded by $owner/$repoSlug#${oldest.number}"))
+      _ <- deleteBranch(pullRequest)
+      _ <- IO(log.info(s"Removed branch ${pullRequest.base.map(_.ref)} from $owner/$repoSlug"))
+    } yield NotOldest(oldest, pullRequest).asLeft[PullRequestDraft]
+
+    // TODO: Update pull request description?
```
Comment (dcsobral) [inline core/src/main/scala/slamdata/AutoBump.scala]: @djspiewak Feedback on this, please. I'm wondering where I'm going to do it -- at the point I made a decision on create/update, it is too soon since I don't know if `push` will succeed.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+   */
+  def getOldestAutoBumpPullRequest: IO[Option[PullRequestDraft]] = {
+    getPullRequests
+      .evalFilter(pullRequest => getLabels(pullRequest.number).map(labels => isAutoBump(pullRequest, labels)))
+      .head
+      .compile.toList
+      .map(_.headOption)
+  }
+
+  // TODO: use Last-Modified/If-Modified-Since
```
Comment (dcsobral) [inline core/src/main/scala/slamdata/AutoBump.scala]: Remove TODO.

File: build.sbt [hunk 8]
```diff
@@ -6,12 +6,14 @@ ThisBuild / scalaVersion := "2.12.10"
 lazy val root = project
   .in(file("."))
   .aggregate(core, artifact, plugin)
+  .settings(name := "sbt-precog-root")
```
Comment (djspiewak) [inline build.sbt:9]: Worth noting that we *may* have some old infrastructure which hardcodes `root` for the root project. I think I replaced all of that with `/` but I may have missed something.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.sys.process._
+import scala.util.Try
+import scala.util.matching.Regex
+
+object AutoBump {
+  sealed trait Warnings {
+    def warn(log: Logger): IO[Unit]
+  }
+  case object UpdateError extends Warnings {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Random style note: we usually put members of an ADT under its companion object. So this would be in `object Warnings`, for example.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+
+import scala.collection.immutable.Seq
+import scala.collection.mutable
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.sys.process._
+import scala.util.Try
+import scala.util.matching.Regex
+
+object AutoBump {
+  sealed trait Warnings {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: When encoding an ADT, it's usually best to make the parent type `extend Product with Serializable` so that your type errors don't look horrible.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+      log.warn("you should check for a stuck trickle PR on that repository")
+    }
+  }
+  case object PushError extends Warnings {
+    def warn(log: Logger): IO[Unit] = IO {
+      log.warn("git-push exited with error")
+      log.warn("this usually means some other repository updated the pull request before this one")
+    }
+  }
+  case class NotOldest(oldest: PullRequestDraft, draft: PullRequestDraft) extends Warnings {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: All case classes should be `final`. Always.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+    val fromString: Map[String, ChangeLabel] = values.map(change => change.label -> change).toMap
+    val labelPattern: Regex = values.mkString("|").r
+
+    def apply(label: String): Option[ChangeLabel] = {
+      fromString.get(label)
+    }
+
+    def unapply(arg: String): Option[ChangeLabel] = labelPattern.findFirstIn(arg).flatMap(apply)
+  }
+  case object RevisionChange extends ChangeLabel("version: revision")
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: One of the advantages to putting things into companion objects is you can simplify these names. Rather than `RevisionChange`, you'll get just `Revision`, inside of `ChangeLabel`.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+  case object FeatureChange extends ChangeLabel("version: feature")
+  case object BreakingChange extends ChangeLabel("version: breaking")
+
+  implicit class Unpack[F[_]: Monad, A](response: F[GHResponse[A]]) {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Minor nit: `UnpackSyntax` is probably slightly more descriptive

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+    def unpack: F[Either[GHException, A]] = EitherT(response).map(_.result).value
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala:93]: Why not just `response.map(_.map(_.result))`?

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+  }
+
+  val AutoBumpLabel = ":robot:"
+  val PullRequestFilters: List[PRFilter] = List(PRFilterOpen, PRFilterSortCreated, PRFilterOrderAsc, PRFilterBase("master"))
+  val LinkRelation: Regex = """<(.*?)>; rel="(\w+)"""".r
+  val LabelPattern: Regex = "version: \\w+".r
+
+  def autoBumpCommitTitle(author: String) = s"Applied dependency updates by $author"
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Minor nit: public members should have explicitly declared types.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+
+  def autoPage[F[_]: Sync, T](first: Pagination)
+                             (call: Pagination => F[Either[GHException, GHResult[List[T]]]]): Stream[F, T] = {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Style nit! This is incorrectly formatted. Actually, most of your long signatures are.

We generally wrap signatures by wrapping all blocks to a double-indented context, *including* the return type, which we wrap before the `:`. Thus, your signature should be the following:

```suggestion
  def autoPage[F[_]: Sync, T](
      first: Pagination)(
      call: Pagination => F[Either[GHException, GHResult[List[T]]]])
      : Stream[F, T] = {
```

Basically as a rule of thumb, if you're vertically aligning *anything* other than sbt's dependency `%` operators, it's not conventional.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+  }
+
+  def nextPage(relations: Map[String, (Int, Int)]): Option[Pagination] = {
+    relations.get("next").map((Pagination.apply _).tupled)
+  }
+
+  /** Decodes github's "Link" header into a map */
+  def getRelations(headers: Map[String, String]): Map[String, (Int, Int)] = {
+    val relations = for {
+      header <- headers.map { case (k, v) => k.toLowerCase -> v }.collect { case ("link", header) => header }
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: More random style nits… *Either* use `.` and `(`, or use <kbd>space</kbd> and `{`, but not a mixture of the two. So the above should be the following if you want to keep it one line:

```suggestion
      header <- headers map { case (k, v) => k.toLowerCase -> v } collect { case ("link", header) => header }
```

Or you can break it up into two lines for readability, but in either case you're going to end up not using `.`.

The above actually nicely illustrates *why* this rule is something you should adhere to: depending on syntactic context, `}.` doesn't do what you think it does. In this case, it *is* doing the "right" thing, but in other contexts it won't, and the rules around it are very subtle. Not quite as bad as in Ruby, but still bad. Just avoid it.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+   * Finds primary open autobump pull request, if one exists.
+   *
+   * @return Oldest autobump pull request, by order of creation
+   */
+  def getOldestAutoBumpPullRequest: IO[Option[PullRequestDraft]] = {
+    getPullRequests
+      .evalFilter(pullRequest => getLabels(pullRequest.number).map(labels => isAutoBump(pullRequest, labels)))
+      .head
+      .compile.toList
+      .map(_.headOption)
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]:
```suggestion
      .compile.last
```

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+  }
+
+  // TODO: use Last-Modified/If-Modified-Since
+  def getPullRequests: Stream[IO, PullRequestDraft] = {
+    autoPage(Pagination(1, 100)) { pagination =>
+      github.pullRequests.listDraftPullRequests(owner, repoSlug, PullRequestFilters, Some(pagination))
+    }
+  }
+
+  // FIXME: current github4s api does not support Some(pagination)
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala:183]: What does this `FIXME` mean? Is this something wrong *here* or in github4s?

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 20]
```diff
@@ -0,0 +1,369 @@
...
+    github.issues.listLabels(owner, repoSlug, pr).rethrow.map(_.result)
+  }
+
+  def draftPullRequest(authorRepository: String, branchName: String, changes: String): IO[PullRequestDraft] = {
+    github
+      .pullRequests
+      .draftPullRequest(
+        owner,
+        repoSlug,
+        NewPullRequestData(autoBumpCommitTitle(authorRepository),
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Style nit: put `autoBumpCommitTitle(authorRepository)` on the next line. Basically if you're wrapping actuals, either wrap all or none.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 21]
```diff
@@ -0,0 +1,369 @@
...
+        updateResult <- run(s"$sbt update", log, merge = true, Some(dirFile)).attempt
+      } yield updateResult.bimap(_ => UpdateError, _ => (dirFile, branchName, oldestPullRequest, changes, label))
+    }
+
+    // TODO: add changes to commit message?
+    def tryCommit(dirFile: File): IO[Either[Warnings, Unit]] = {
+      for {
+        _ <- run(s"git add .", log, merge = true, workingDir = Some(dirFile))
+        result <- run(
+          Seq("git", "commit", "-m", autoBumpCommitTitle(authorRepository)),
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: So what we *used* to do here is build a mini changelog from the origin repo(s). Basically, grab the origin repo, do a `git log vXXX..HEAD`, filter to the merge commits descending from each other (I can't remember exactly how this works, but git has a flag for it) and take the commit messages (which are derived from pull request titles by `sdmerge`).

I'd like to revive this at some point, but it's harder with trickle since you'd need to go clone *those* repos, and also the changelog thing wasn't really perfect anyway. This is fine for now, and we can refine it over time.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 22]
```diff
@@ -0,0 +1,369 @@
...
+      _ <- IO(log.info(s"Closed $owner/$repoSlug#${pullRequest.number}, preceded by $owner/$repoSlug#${oldest.number}"))
+      _ <- deleteBranch(pullRequest)
+      _ <- IO(log.info(s"Removed branch ${pullRequest.base.map(_.ref)} from $owner/$repoSlug"))
+    } yield NotOldest(oldest, pullRequest).asLeft[PullRequestDraft]
+
+    // TODO: Update pull request description?
+    def createOrUpdatePullRequest(branchName: String,
+                                  changes: List[String],
+                                  changeLabel: ChangeLabel,
+                                  maybePullRequest: Option[PullRequestDraft]): IO[Either[Warnings, PullRequestDraft]] = {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Another style example:

```suggestion
    def createOrUpdatePullRequest(
        branchName: String,
        changes: List[String],
        changeLabel: ChangeLabel,
        maybePullRequest: Option[PullRequestDraft])
        : IO[Either[Warnings, PullRequestDraft]] = {
```

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 23]
```diff
@@ -0,0 +1,369 @@
...
+      log.warn("this usually means two or more repositories finished build at the same time,")
+      log.warn("and some other repository beat this one to pull request creation.")
+    }
+  }
+
+  sealed abstract class ChangeLabel(val label: String)
+  object ChangeLabel {
+    val values: List[ChangeLabel] = List(RevisionChange, FeatureChange, BreakingChange)
+    val order: Map[ChangeLabel, Int] = values.zipWithIndex.toMap
+    implicit val ordering: Ordering[ChangeLabel] = Ordering.by(order)
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Minor aside: we have Cats in scope here. Rather than using `Ordering`, it's probably better to just use `Order`. I'm not super-fussed about it either way.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 6 above)
Comment (djspiewak) [reply core/src/main/scala/slamdata/AutoBump.scala]: If the push *doesn't* succeed, then you just drop it. If we were to make changes to the existing PR, then we would probably do them *after* the push.

I think it's fine to not do anything to the existing PR description. We aren't putting information into it right now anyway, so there's nothing really to update. Eventually, when we *do* add more information to the autobump PRs, we will of course want to keep them updated, but right now neither case is true.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 4 above)
Comment (djspiewak) [reply core/src/main/scala/slamdata/AutoBump.scala]: In general, you want to push `IO` as far out to the leaves as possible. In this case, that would mean returning `IO` (i.e. *not* using `unsafe` anything) all the way to the setting/task definitions, where you would of course need to use the `unsafe` function to run it. There isn't *that* much of a difference in terms of testability, but it does help a lot in terms of reasoning when and where effects are run.

So tldr, I do think returning `IO` here would be better.

File: core/src/main/scala/slamdata/AutoBump.scala [hunk 24]
```diff
@@ -0,0 +1,369 @@
...
+                  workingDir: Option[File],
+                  env: (String, String)*): IO[List[String]] = {
+    val lines = mutable.Buffer[String]()
+    val processLogger = getProcessLogger(log, merge, lines)
+    run_?(command, log, processLogger, workingDir, env: _*)
+      .ensureOr(res => new RuntimeException(s"${command.take(2).mkString(" ")} exit code $res"))(_ == 0)
+      .as(lines.toList)
+  }
+
+  private def run_?(command: Seq[String],
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/AutoBump.scala]: Suffix operators on alphabetical identifiers. Now there's a style I haven't seen [in a long, long time…](https://github.com/lift/framework)

File: core/src/main/scala/slamdata/SbtPrecogBase.scala [hunk 25]
```diff
@@ -456,131 +442,58 @@ abstract class SbtPrecogBase extends AutoPlugin {
...
+              case Some(currentRevision) =>
+                val currentVersion = VersionNumber(currentRevision)
+                val newVersion = VersionNumber(newRevision)
+                val testRevision = VersionNumber.SecondSegment.isCompatible(currentVersion, newVersion)
+                val testBreaking = !VersionNumber.SemVer.isCompatible(currentVersion, newVersion)
+                isRevision &&= testRevision
+                isBreaking ||= testBreaking
+                log.info(s"Updated ${getChange(testRevision, testBreaking)} $dependencyRepository $currentVersion -> $newRevision")
+              case None                  =>
+                // TODO: use scalafix to change build.sbt
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/SbtPrecogBase.scala:534]: That would be cool…

File: core/src/main/scala/slamdata/algebras/GitData.scala [hunk 26]
```diff
@@ -0,0 +1,21 @@
+package precog.algebras
+
+import github4s.GithubResponses.GHResponse
+import github4s.domain.Ref
+
+trait GitData[F[_]] extends github4s.algebras.GitData[F] {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/algebras/GitData.scala:22]: Trivia note! This is the very first time I've ever worked with or reviewed code that is *actually* using a finally tagless encoding.

File: core/src/main/scala/slamdata/domain/PullRequestDraft.scala [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package precog.domain
+
+import github4s.domain.{PullRequest, PullRequestBase, User}
+
+case class PullRequestDraft(id: Int,
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/domain/PullRequestDraft.scala:21]: Should be `final`. Also class formals should be formatted analogously to method formals: when split across lines, double-indent rather than vertically aligning, with `extends` and `with` being similarly double-indented and on their own line(s).

File: core/src/main/scala/slamdata/domain/PullRequestDraft.scala [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+                            created_at: String,
+                            updated_at: Option[String],
+                            closed_at: Option[String],
+                            merged_at: Option[String],
+                            merge_commit_sha: Option[String],
+                            base: Option[PullRequestBase],
+                            head: Option[PullRequestBase],
+                            user: Option[User],
+                            assignee: Option[User]
+                           )
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/domain/PullRequestDraft.scala:39]: Another random style nit: we usually don't favor dangling `)` in this fashion. Instead, we put it at the end of the previous line. This distinguishes the scope a bit more from `}` (which we *do* push to the next line!), where the newline inference semantics are different.

File: core/src/main/scala/slamdata/domain/PullRequestDraft.scala [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+
+object PullRequestDraft {
+//  implicit def toPullRequest(pr: PullRequestDraft): PullRequest = PullRequest(
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/domain/PullRequestDraft.scala]: A rare valid use of implicit coercion, btw.

File: core/src/main/scala/slamdata/domain/PullRequestDraft.scala [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+sealed trait DraftPullRequest {
+  def head: String
+  def base: String
+  def maintainer_can_modify: Option[Boolean]
+  def draft: Boolean
+}
+
+case class DraftPullRequestData(title: String,
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/domain/PullRequestDraft.scala]: Moar final…

File: core/src/main/scala/slamdata/interpreters/GitDataInterpreter.scala [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+import cats.effect.Sync
+import cats.implicits._
+import github4s.GithubResponses.GHResponse
+import github4s.domain.{Ref, RefAuthor, RefCommit, RefInfo, Tag, TreeData, TreeResult}
+import github4s.http.HttpClient
+import precog.algebras.GitData
+
+class GitDataInterpreter [F[_] : Sync](impl: github4s.algebras.GitData[F])
+                                      (implicit client: HttpClient[F], accessToken: Option[String])
+    extends GitData[F] {
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/interpreters/GitDataInterpreter.scala]:
```suggestion
class GitDataInterpreter[F[_]: Sync](
    impl: github4s.algebras.GitData[F])(
    implicit client: HttpClient[F],
    accessToken: Option[String])
    extends GitData[F] {
```

Why is the access token `implicit`? Also the client? But really especially the access token.

File: core/src/main/scala/slamdata/interpreters/PullRequestsInterpreter.scala [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+
+object PullRequestsInterpreter {
+  import io.circe._
+  import io.circe.generic.auto._
+  import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
+  import io.circe.syntax._
+  import github4s.Decoders._
+
+  /** Representation of the json data github graphql queries expect. */
+  case class GithubQuery[A <: GraphQLQuery](query: A, variables: A#Variables)
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/interpreters/PullRequestsInterpreter.scala]: First off, `final`. Second, this is a surprisingly tricky type particularly taken into context with the implicits in the companion object. Is it actually strictly necessary?

**Review by djspiewak — COMMENTED** [review]
Mostly a ton of style nits and Scala best practice points. Nothing too major. I would be fine merging this as-is, in fact. Nice work.

Regarding your major points on PR descriptions, I definitely think we should add more detail to them down the line, but right now we should just get this out the door and refine it later. It would be really nice to be able to see at a glance (when reviewing trickle PRs) what is going on, similar to what we had before with autobumps, but it's workable without it.

File: build.sbt (hunk 8 above)
Comment (dcsobral) [reply build.sbt:9]: This is just for sbt-precog. I had forgotten about it, tbh. I just did it because IntelliJ is annoying -- every time it reloads the project, it replaces the project name with root's name, so I end up trying to pick which project called "root" I want to load.

Comment (dcsobral) [reply build.sbt:9]: Wait... what do you mean by replacing with `/`?

**Comment by dcsobral** [comment]
> Regarding your major points on PR descriptions, I definitely think we should add more detail to them down the line, but right now we should just get this out the door and refine it later. It would be really nice to be able to see at a glance (when reviewing trickle PRs) what is going on, similar to what we had before with autobumps, but it's workable without it.

We do tag every release, don't we? I just realized getting the commit logs is easier than I thought. I was too focused on how the old autobump worked to realize I can simply use github4s to retrieve the logs without checking out the repositories.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 11 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: And it is rather annoying that they aren't. :/

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 12 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: True. Curiously, nowhere in the code actually directly refers to these. I struggled on whether to make the code in `SbtPrecogBase` that decides what the label should be directly refer to these or not. I left this as is to avoid circular references, though both ADTs really ought to be moved out of `AutoBump`.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 14 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala:93]: I dislike nesting maps. I get lost...

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 15 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: And IntelliJ annoys me every time I don't do it, so I don't understand why it didn't do so here. That said, this used to be `val AutoBumpCommitTitle`, in which case we _should_ leave the type out to get constant semantics. That's *got* to be the most annoying feature of Scala in my list.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 16 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: Ugh, I hate dangling return types. :(

Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: If we had a scalastyle configs, IntelliJ would have pointed that stuff out to me.

File: build.sbt (hunk 8 above)
Comment (djspiewak) [reply build.sbt:9]: In sbt commands, `/` is an alias for the root project, whatever it may be named.

**Comment by djspiewak** [comment]
> We do tag every release, don't we? I just realized getting the commit logs is easier than I thought. I was too focused on how the old autobump worked to realize I can simply use github4s to retrieve the logs without checking out the repositories.

Ah that's a good point. Yes, we do, barring bugs. Sometimes weird things happen and the tags don't get created and we shouldn't fail if that's the case.

If it's easy to compute changelogs, go for it. :-) I'm also fine just doing that in a follow-up PR.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 17 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: I'm well familiar with parenthesis vs braces in Scala. I thought you didn't like operator notation on non-symbolic methods.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 19 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala:183]: Both. It is wrong here because github4s does not support pagination in this particular API. In fact, I filed an issue there, and it's fixed on master.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 21 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: It's even harder because, the way it was done, we knew for sure what the previous version was. It's trickier now, particularly with autobump PR updates. As for checking out... we don't need to do that. We can query github.

Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: OH, another thing... originally, every bump corresponded to a single repository version bump. Now we are combining bumps from multiple repositories.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 6 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: Actually, we are putting information there now. Just what the revision update is, but... it's going there.

File: build.sbt (hunk 8 above)
Comment (dcsobral) [reply build.sbt:9]: How can it be an alias for root _and_ a separator for scopes?

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 18 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: Last? Steam is returning a list of prs <<pr1, pr2, pr3, ..., prn>>, where pr1 is the oldest and prn is the newest, and I want the oldest, which is what the code was doing. `last` seems to do very much the opposite,no?

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 23 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: My mind was about to go into shutdown, and I'm not familiar enough with Cats yet. Hell, even dragging out the details of `Ordering` was a major ordeal at that point.

File: core/src/main/scala/slamdata/AutoBump.scala (hunk 24 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/AutoBump.scala]: Nothing else occurred to me, I'm afraid. It felt wrong and icky, but also understandable. I'm still a bit disgruntled that `!` and `?` cannot appear as suffixes to alphabetical identifiers, ten years later. :)

File: core/src/main/scala/slamdata/interpreters/GitDataInterpreter.scala (hunk 31 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/interpreters/GitDataInterpreter.scala]: I've no idea. I just copy & pasted the declaration from github4s, taking it on faith that it made sense.

Comment (dcsobral) [reply core/src/main/scala/slamdata/interpreters/GitDataInterpreter.scala]: I guess just to reduce duplication initializing all these classes on `Github`. <sigh>

File: build.sbt (hunk 8 above)
Comment (djspiewak) [reply build.sbt:9]: It's only an alias for root in the shell, and I *think* only in the `project` command. So you can run `project /`, but I don't think you can do something like `//test` and run `test` in root (maybe you can… dunno).

**Comment by djspiewak** [comment]
Needs merge from master btw

File: core/src/main/scala/slamdata/Github.scala [hunk 33]
```diff
@@ -0,0 +1,63 @@
...
+      new github4s.interpreters.PullRequestsInterpreter[F],
+      client,
+      accessToken)
+  val users: Users[F]                    = new UsersInterpreter[F]
+  val repos: Repositories[F]             = new RepositoriesInterpreter[F]
+  val auth: Auth[F]                      = new AuthInterpreter[F]
+  val gists: Gists[F]                    = new GistsInterpreter[F]
+  val issues: Issues[F]                  = new IssuesInterpreter[F]
+  val activities: Activities[F]          = new ActivitiesInterpreter[F]
+  val gitData: GitData[F]                = new GitDataInterpreter[F](
```
Comment (djspiewak) [inline core/src/main/scala/slamdata/Github.scala:46]: IntelliJ's auto-formatter kinda went nuts here…

File: core/src/test/scala/precog/RunnerSpec.scala [hunk 34]
```diff
@@ -0,0 +1,164 @@
...
+  val log = sbt.util.LogExchange.logger("test")
+
+  "Fluent builders" should {
+    "be equivalent to case class build" in {
+      val runner = Runner[IO](log)
+      runner === Runner[IO](log, false, None, Map.empty, Seq.empty)
+      runner.stderrToStdout === Runner[IO](log, merge = true)
+      runner.cd(file("./target")) === Runner[IO](log, workingDir = Some(file("./target")))
+      runner.withEnv("X" -> "Y", "W" -> "Z") === Runner[IO](log, env = Map("X" -> "Y", "W" -> "Z"))
+      runner.hide("password") === Runner[IO](log, hide = Seq("password"))
```
Comment (djspiewak) [inline core/src/test/scala/precog/RunnerSpec.scala:35]: We strongly prefer a `mustEqual` style due to the fact that `===` is a Cats equality operator.

File: core/src/test/scala/slamdata/AutoBumpTest.scala [hunk 35]
```diff
@@ -0,0 +1,45 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package precog
+
+import org.specs2.mutable.Specification
+
+class AutoBumpTest extends Specification {
```
Comment (djspiewak) [inline core/src/test/scala/slamdata/AutoBumpTest.scala:21]: `AutoBumpSpec`

**Comment by djspiewak** [comment]
LGTM. Merging as-is. Any other changes we can address in a follow-up PR if necessary

File: core/src/main/scala/slamdata/Github.scala (hunk 33 above)
Comment (dcsobral) [reply core/src/main/scala/slamdata/Github.scala:46]: Actually, this file was copied from github4s, and I just didn't bother reformatting it.

File: core/src/test/scala/slamdata/AutoBumpTest.scala (hunk 35 above)
Comment (dcsobral) [reply core/src/test/scala/slamdata/AutoBumpTest.scala:21]: Yeah. I'm letting IntelliJ create it, and it has a weird default. Well, perhaps not _weird_, just not well suited to Specs.

File: core/src/main/scala/slamdata/Github.scala (hunk 33 above)
Comment (djspiewak) [reply core/src/main/scala/slamdata/Github.scala:46]: Gotcha
